okta/okta.go

```diff
@@ -46,26 +52,40 @@ type resource struct {
 	client *Client
 }
 
-func NewClient(config *Config, httpClient *http.Client, cacheManager cache.Cache) *Client {
-	if config == nil {
-		config = NewConfig()
+func NewClient(ctx context.Context, conf ...ConfigSetter) (*Client, error) {
+	config := &config{}
+
+	setConfigDefaults(config)
+	config = readConfigFromSystem(*config)
+	config = readConfigFromApplication(*config)
+	config = readConfigFromEnvironment(*config)
+
+	for _, confSetter := range conf {
+		confSetter(config)
 	}
 
 	var oktaCache cache.Cache
 	if !config.Okta.Client.Cache.Enabled {
 		oktaCache = cache.NewNoOpCache()
 	} else {
-		if cacheManager == nil {
+		if config.CacheManager == nil {
 			oktaCache = cache.NewGoCache(config.Okta.Client.Cache.DefaultTtl,
 				config.Okta.Client.Cache.DefaultTti)
 		} else {
-			oktaCache = cacheManager
+			oktaCache = config.CacheManager
 		}
 	}
 
+	config.CacheManager = oktaCache
+
+	config, err := validateConfig(config)
+	if err != nil {
+		panic(err)
```

Comment:
I assume this is similar to C++ exception. If it is, then probably we don't need to return tuples like `(*Client, error)`, as looks like there will be no statements like: `return nil, error`

Reply:
Validate config does `return nil, err` if the config is invalid

Reply:
Just curios whether user is able to recover from the error during runtime? Or it is end-of-life event?

Reply:
Yes, you can recover.  


https://blog.golang.org/defer-panic-and-recover